Fast Kalman Gyro Filter: Implementation and parameter groups only

fujin opened the pull request:
As per https://github.com/betaflight/betaflight/pull/4887#issuecomment-354937159 here we have @rs2k's Fast Kalman implementation and parameter group to suit. Note that it is disabled by default. I also incorporated all of the feedback from the previous PR.

Signed-off-by: AJ Christensen <aj@junglistheavy.industries>

fujin commented:
Hmm, it doesn't look like this is fitting in the `FLASH` region :(

```/home/travis/build/betaflight/betaflight/tools/gcc-arm-none-eabi-7-2017-q4-major/bin/../lib/gcc/arm-none-eabi/7.2.1/../../../../arm-none-eabi/bin/ld: obj/main/betaflight_FRSKYF3.elf section `.data' will not fit in region `FLASH'
/home/travis/build/betaflight/betaflight/tools/gcc-arm-none-eabi-7-2017-q4-major/bin/../lib/gcc/arm-none-eabi/7.2.1/../../../../arm-none-eabi/bin/ld: region `FLASH' overflowed by 73 bytes```

Suggestions? Make it #DEFINE-able?

martinbudden commented on src/main/common/filter.c:
```diff
@@ -302,3 +302,33 @@ float firFilterDenoiseUpdate(firFilterDenoise_t *filter, float input)
 }
+
+// Fast two-state Kalman
+void fastKalmanInit(fastKalman_t *filter, float q, float r, float p, float initialValue)
```
`initialValue` not required, since it is only ever set to zero. That should save a few bytes of ROM.

martinbudden commented on src/main/interface/settings.c:
```diff
@@ -356,6 +356,10 @@ const clivalue_t valueTable[] = {
+    { "gyro_kalman_q",              VAR_UINT16 | MASTER_VALUE, .config.minmax = { 0, 16000 }, PG_GYRO_CONFIG, offsetof(gyroConfig_t, gyro_kalman_q) },
+    { "gyro_kalman_r",              VAR_UINT16 | MASTER_VALUE, .config.minmax = { 0, 16000 }, PG_GYRO_CONFIG, offsetof(gyroConfig_t, gyro_kalman_r) },
+    { "gyro_kalman_p",              VAR_UINT16 | MASTER_VALUE, .config.minmax = { 0, 16000 }, PG_GYRO_CONFIG, offsetof(gyroConfig_t, gyro_kalman_p) },
+    { "gyro_kalman_enable",         VAR_UINT8  | MASTER_VALUE | MODE_LOOKUP, .config.lookup = { TABLE_OFF_ON }, PG_GYRO_CONFIG, offsetof(gyroConfig_t, gyro_kalman_enable) },
```
Standard practice is to disable filters by setting one of their parameters to zero, ie set `gyro_kalman_p` to zero.

As well as being standard practice, it should save a few bytes of ROM.

martinbudden commented:
@fujin , if you make both my suggested changes, you might just save enough bytes to fit in the ROM.

fujin commented:
@martinbudden thanks for reviewing - much appreciated. I have addressed your feedback but it still doesn't quite fit in the FRSKYF3 - overflows by 41 bytes.

```Linking FRSKYF3
/Users/aj/dev/betaflight/tools/gcc-arm-none-eabi-7-2017-q4-major/bin/../lib/gcc/arm-none-eabi/7.2.1/../../../../arm-none-eabi/bin/ld: obj/main/betaflight_FRSKYF3.elf section `.data' will not fit in region `FLASH'
/Users/aj/dev/betaflight/tools/gcc-arm-none-eabi-7-2017-q4-major/bin/../lib/gcc/arm-none-eabi/7.2.1/../../../../arm-none-eabi/bin/ld: region `FLASH' overflowed by 41 bytes
collect2: error: ld returned 1 exit status
make[1]: *** [obj/main/betaflight_FRSKYF3.elf] Error 1
make: *** [hex] Error 2```

martinbudden commented:
I've found some ROM saving elsewhere. I hope to get a PR in soon.

McGiverGim commented:
What is the idea about this filter? Use it in place of lpf? in place of notch? Add it to the others in very noise setups?

Seeing the code, it seems that it can be added to the others filters, but I suppose that the idea is to use it in place of one of the others (if the user decide that it works better in his setup). 

How to configure it? I have seen some "default initial" values in the original PR, but I don't know if there's a some info in the wiki or in other place about this.

ledvinap commented:
@McGiverGim : after initial transient, this filter converges to simple RC with coefficient `alpha = 2*k`:
`p` is dependent only on `p` and constant `q`, `r`, it will quickly converge to constant value
`k = p / (p + r)`
`x = x + 2*k * (input - x)` // k * (input - x) is added second time in predict step

Compare this to https://en.wikipedia.org/wiki/Low-pass_filter#Discrete-time_realization

This is expected result - Kalman filter is calculating optimal gain.
This may be interesting to calculate optimal cutoff frequency if accurate process noise is available, but guessing reasonable noise values is hard part of Kalman design ..
Also note that Kalman expects independent Gaussian noises, actual noise shapes in quadcopters are far from gaussian...

conkerkh commented:
Just an idea but most likely adding angle estimation from accelerometer and gyro would give more accurate results on dps as well. I'm far from expert in filtering but afaik the idea is that you could have both readings influence another hence any noise on gyro could be also filtered out by accelerometer readings. Probably only thing that might worry is step response but with appropriate tuning it should be ok.

denovich commented:
@ledvinap The non-gausian elements of the noise are why a simple kalman alone isn't the answer. Another filter is needed to address the resonances/motor noise/etc.  

@McGiverGim https://www.rcgroups.com/forums/showthread.php?3009636-BF-K-Testing-the-Fast-Kalman-filter-on-Betaflight has the details...  

@conkerkh pretty sure @rs2k has already tried this... if you think these gyro is noisy, have a look at the accelerometers, they are even worse.

ledvinap commented:
Sorry, I overlooked one thing in analysis above (thanks rtlopez to point notice this on RCGroups):
Output is taken between  applying `k * (input - x)` twice - this acts as 2 point averaging (or 2 point FIR lowpass) filter, applies in series with RC low-pass (ordering is arbitrary due to LTI).
If this short FIR filter improves performance, it may be interesting to investigate FIR filtering for high gyro rates ...

fujin commented:
I have a local changeset which contains the minor modifications to support being rebased to the following PR's submitted by @martinbudden today:

- https://github.com/betaflight/betaflight/pull/4892
- https://github.com/betaflight/betaflight/pull/4893
- https://github.com/betaflight/betaflight/pull/4894 

Notably, the `(filter_t *)` casts for the fastKalmanUpdate calls. With this code, it fits on the FRSKYF3 ROM without overflow, it is at least dependent on #4894.

Is the standard practice to wait until the others are merged - or I can mark this as dependent of them in some way besides mentioning it here? it doesn't seem right to include the rebased changes all in this one. I also tagged him in Slack but wanted to make record of it here.

martinbudden commented:
Standard practice is to wait for the other PRs to be merged in, and then rebase your PR on that. The other PRs are just waiting review before they can be merged.

fujin commented:
Thanks @martinbudden, understood.

fujin commented:
In https://github.com/betaflight/betaflight/pull/4890/commits/e04d3cf1a2aa4967c049faf7cc81a59788f5cb37, I have introduced a new `#define USE_GYRO_FAST_KALMAN`, which can be used to conditionally include the FKF on (or exclude it from) certain targets.

All targets, except for the `OMNIBUS`, appear to have enough ROM available. On this target, I have `#undef` the define. Without:
```
Linking OMNIBUS 
/home/aj/dev/betaflight/tools/gcc-arm-none-eabi-7-2017-q4-major/bin/../lib/gcc/arm-none-eabi/7.2.1/../../../../arm-none-eabi/bin/ld: obj/main/betaflight_OMNIBUS.elf section `.data' will not fit in region `FLASH'
/home/aj/dev/betaflight/tools/gcc-arm-none-eabi-7-2017-q4-major/bin/../lib/gcc/arm-none-eabi/7.2.1/../../../../arm-none-eabi/bin/ld: region `FLASH' overflowed by 356 bytes
collect2: error: ld returned 1 exit status
Makefile:285: recipe for target 'obj/main/betaflight_OMNIBUS.elf' failed
make[1]: *** [obj/main/betaflight_OMNIBUS.elf] Error 1
```

I could use some help tracking down what the `OMNIBUS` target in particular has in terms of features or peripherals which have this huge FLASH memory footprint. I'm not that keen on the `#define`-based approach - but this may be acceptable for now.

martinbudden commented on src/main/common/filter.c:
```diff
@@ -302,3 +302,35 @@ float firFilterDenoiseUpdate(firFilterDenoise_t *filter, float input)
     }
 }
+
+#if defined(USE_GYRO_FAST_KALMAN)
```
This is not required, since compiler will not include this if it is not used.

martinbudden commented on src/main/common/filter.h:
```diff
@@ -53,6 +53,17 @@ typedef struct firFilterDenoise_s {
     float state[MAX_FIR_DENOISE_WINDOW_SIZE];
 } firFilterDenoise_t;
 
+#if defined(USE_GYRO_FAST_KALMAN)
```
This is not required.

martinbudden commented on src/main/common/filter.h:
```diff
@@ -87,6 +98,11 @@ float biquadFilterApplyDF1(biquadFilter_t *filter, float input);
 float biquadFilterApply(biquadFilter_t *filter, float input);
 float filterGetNotchQ(uint16_t centerFreq, uint16_t cutoff);
 
+#if defined(USE_GYRO_FAST_KALMAN)
```
This is not required.

martinbudden commented on src/main/target/common_fc_pre.h:
```diff
@@ -43,13 +43,15 @@
 #define USE_DSHOT
 #define USE_DSHOT_DMAR
 #define USE_GYRO_DATA_ANALYSE
+#define USE_GYRO_FAST_KALMAN
```
`USE_GYRO_FAST_KALMAN` should be `#define`d based on ROM size, not processor type.

fujin commented:
thanks @martinbudden, that's a smarter way to handle it.. I'm sorry for not noticing :)

handled in https://github.com/betaflight/betaflight/pull/4890/commits/6e6aafe6d54ca070c52ad4ba8a670a1ed63e2b57 - cheers!

martinbudden commented:
@mikeller , I think this can be merged now. What do you think?

conkerkh commented:
Personally I think it's confusing that filter is enabled once q and r params are != 0, this is weird a bit. I would add some gyro_enable_kalman settings and set q and r defaults when reseting eeprom. This would be more intuitive. As it was (is?) in rs2k fork

ledvinap commented:
@martinbudden : Blockchain should be added too. Blackbox logs are implemented too efficiently ...

conkerkh commented on src/main/common/filter.c:
```diff
@@ -302,3 +302,33 @@ float firFilterDenoiseUpdate(firFilterDenoise_t *filter, float input)
     }
 }
+
+// Fast two-state Kalman
```
Is this really two-state kalman? We are only estimating one state that is gyro rate

martinbudden commented:
@ledvinap , sorry, I don't understand your comment.

ledvinap commented:
@martinbudden : Sorry, bad joke ... 
Kalman, as used in this PR, is only buzzword.

jflyper commented:
So, what happened to `gyro_enable_kalman`? Did it ever get in?

conkerkh commented:
@jflyper Doesn’t look like it:/

martinbudden commented:
@ledvinap , while I agree that this filter may not give much benefit compared with the current filters, the cost of including it is low and it will enable pilots to test it.

mikeller commented:
@jflyper, @conkerkh: I was humming and harring about that one. But whereever applicable in Betaflight we use setting of a control parameter to 0 to disable a function, and in some cases (notch filters) this is done for a group of parameters, so to me the current implementation seems to be most aligned with the established pattern

seaweeduk commented:
I think this filter should be called Kalyn filter rather than Kalman filter

ledvinap commented:
@martinbudden : I will make BF code a bit more complicated. It will confuse new programmers in future. It will waste time in support forums that could be used for actual problems. It may create placebo effect, spreading further.

And it may mask the fact that short FIR could be very helpful at high gyro rates.

DanNixon commented:
@conkerkh https://github.com/betaflight/betaflight/pull/4890#pullrequestreview-86313055 explains why `gyro_enable_kalman` was removed.

denovich commented:
It may do all thise things. 
It may also provide superior flight performance compared with existing filtering options. 
The later having already been enthusiastically reported by many users already.

ledvinap commented:
@denovich : Using RC+2-point averaging will provide exactly same results (considerably faster), with single, meaningful parameter (cutoff frequency)

DieHertz commented:
>The later having already been enthusiastically reported by many users already.
That's what @ledvinap mentioned by placebo effect.

ledvinap commented:
@DieHertz : 2-pole FIR filter (averaging 2 values) has high attenuation at half of sample frequency. This may be useful and may be behind good results.. It is worth investigating (but it's hard to investigate it hidden in Kalman update)

conkerkh commented:
I don't see it as a major problem just throw in filter into matlab you gonna instantly how it performs. In 1d with one state estimate it becomes sort of averaging filter, but I think it uses less RAM :) Anyway nice matlab plots would be nice either to confirm it makes a difference or not.

denovich commented:
@DieHertz and @ledvinap attributing it the placebo effect, is nice way of saying all the testers who have been flying this the last few weeks are idiots. I'd like to respond in kind, but I'm failing to find a nice suggest a biologically impossible action.

@ledvinap looking forward to testing your PR with your improved implementation. The considerably faster code should be an even better placebo.

denovich commented:
@conkerkh is matlab easier than flashing BF and flying? Am I grossly underestimating the pilot/tuning skills required to evaluate a filter?

conkerkh commented:
using matlab in this case would be proper scientific way of proving particular thesis... Flashing firmware and flying quad is purely personal experience definitely non scientific, FYI I've flashed my most problematic quad which had never-ending twitches and it's finally flying fine with kalman. But nevertheless AFAIK before notch filter was added there was a lot of discussion and calculations done, so possibly some analysis should be done on Kalman so we have obvious numbers.

DanNixon commented:
@denovich The advantage of a Matlab (or any other) simulation is that it provides an evaluation of filter performance in an isolated manner without influence from other factors.

Test flights cannot be carried out in the same deterministic manner owing to a wide range of factors, the pilot certainly being one of them.

If the reports from pilots that have tested it came from blind A/B testing over a wide range of test flights then I'd be more likely to rule out the placebo effect.

DieHertz commented:
@denovich I prefer to rely on scientific data and carefully crafted tests rather than subjective opinions.

As @DanNixon mentioned, this should be tested with sound and rigorous approaches like A/B testing at the very least, otherwise it's not even sociology, just a preferences poll.

DieHertz commented:
By the way, placebo and nacebo are existing effects and have nothing to do with someone being an idiot, in general.

jflyper commented:
Mob effect accelerates placebo also.

denovich commented:
Yes, but it's not a mob doing the testing.

You are saying the people on our test team, pilots with a huge amount of testing experience, pilots I know and trust, can't be trusted to objectively test a firmware, to read a black box log or to evaluate their own personal experiences.  That is showing a great lack of respect. 

You are also suggesting that I don't know what the word placebo means, and that I don't understand how @ledvinap intended it. Which also shows a lack of respect. It was deliberately used to cast shade on the implementation.

conkerkh commented:
@denovich Software-developers trust the data not opinions so that you understand people approach here. No one is disrespecting anyone.

codecae commented:
If you were driving a car without a speedometer and someone said "you're going 73mph", would you believe them?  You could trust their subjective opinion, but to really know how fast you're going, you need to look at the data.

Same goes for these filters.  Someone's opinion is great for feedback, but to truly know whether or not something is effective, the data is the only real way to know for sure.

fujin commented:
@ledvinap I am happy to help with an implementation of a 2-pole FIR filter if that is beneficial to the project. That way users or researchers will be able to do side-by-side testing or what-have-you. I assume the existing FIR filter type / plumbing may be usable or adapted as well.

@DanNixon @DieHertz @conkerkh @codecae et. al.: I do not possess the understanding (or have a copy of) Matlab/Simulink to do the discussed research there. If someone would like to volunteer or take that on, I would be happy to help in any way that I can. Alternatively, if one can show me how to do the needful with GNU Octave that may be a way forward.

Can we somehow take logs from blackbox and get them into matlab/octave? Does any of this tooling exist (perhaps used for previous filtering implementations?)

timothyswt commented:
I think this conversation went a little off the rails.

The only objective thing here is data. If the data shows better filtering performance or even the same at a lower computational cost, the benefits exist. Obviously external factors cannot be ultimately accounted for until it is fully in the wild. 

@denovich has the data from testing different permutations in various conditions. If you don’t “trust” the data, but have no actual way to refute it other than to argue, one questions the motives of arguing in the first place.

DanNixon commented:
@timothyswt It entirely depends on how the data has been captured, if it is just a series of blackbox logs then what I've previously said still applies. However until said data is available it is a rather moot point.

DieHertz commented:
@timothyswt where is said data? So far there are only impressions.

ledvinap commented:
Here is the math rationale. It is easily possible that I made some error, please check it .. 

Kalman source:
```
     // project the state ahead using acceleration
     filter->x += (filter->x - filter->lastX);
 
     // update last state
     filter->lastX = filter->x;
 
     // prediction update
     filter->p = filter->p + filter->q;
 
     // measurement update
     filter->k = filter->p / (filter->p + filter->r);
     filter->x += filter->k * (input - filter->x);
     filter->p = (1.0f - filter->k) * filter->p;
 
     return filter->x;
```
- `q`       process noise covariance
- `r`       measurement noise covariance
- `p`       estimation error covariance matrix
- `k`       kalman gain
- `x`       state

`q` and `r` are user-specified constant. 

`p` is dependent only on `q` and `r` (and on itself), it is independent of input value. Its value will converge to some value and then stay constant. 

`k` is dependent only on `p` and `r`, so it will converge too.

Using the fact that in steady state `p` and `k` won't change, we can solve value of `k` as function of `q` and `r`. `p` after iteration must stay the same as `p` entering it.
Substitute `t = p + q` (value of `p` after prediction update) and solve the equations (I can't force Wolfram|Alpha to eliminate `t` when solving for 'k', so just solve for `t`):
http://www.wolframalpha.com/input/?i=solve+for+t:+t%3Dp%2BQ,+k%3Dt%2F(t%2BR),+p%3D(1-k)*t

We take positive result (for `k`):
```
k = 2 * sqrt(q) / (sqrt(q + 4 * r) + sqrt(q))
```

From simulation, `k` convergence depends on parameters, when `gyro_kalman_p`==`gyro_kalman_q`, about 100 iterations is enough (84 iterations till `k` is within 1% of final value).

Now with `k` being constant, filter code reduces to:
```
     // project the state ahead using acceleration
     filter->x += (filter->x - filter->lastX);
     // update last state
     filter->lastX = filter->x;
     // measurement update
     filter->x += filter->k * (input - filter->x);
     return filter->x;
```
`filter->x - filter->lastX` is equal to 'measurement update' of last iteration, `filter->k * (input - filter->x)`. This means that measurement update is added twice, but value with half update is returned as value of filter. This is equivalent to adding `2 * filter->k * (input - filter->x);` during filter update and averaging last two values:
```
filter->x += 2 * filter->k * (input - filter->x);
float ret = (filter->x + filter->lastX) / 2;
filter->lastX = filter->x;
return ret;

```

First part is equivalent to RC filter (PT1 in BF):
https://github.com/betaflight/betaflight/blob/master/src/main/common/filter.c#L53

Second part is averaging last two values, for example `firFilterInit2` with coefs `{0.5, 0.5}` 

Another (interesting) possibility is to use biquad filter code to implement RC + FIR2 (implement RC, convolve B with [0.5 0.5]:
```
void biquadFilterInitKalman(biquadFilter_t *filter, float q, float r)
{
  float k = 2 * 2 * sqrt(q) / (sqrt(q + 4 * r) + sqrt(q));  // RC filter k
  filter->b0 = k / 2;
  filter->b1 = k / 2;
  filter->b2 = 0;
  filter->a1 = -(1 - k);
  filter->a2 = 0;
}
```
At least in Octave simulation, this filter is equivalent (after initial transient).

timothyswt commented:
@DanNixon why wouldn't blackbox logs be sufficient? In my own testing, the curve of the noise profiles are smoother than with simple dynamic filtering enabled. The pid loop is simply doing less work overall. How is that data alone not sufficient to prove the benefit of Kalman filtering? Has @denovich and his crew given you prior reason to doubt their methods?

@DieHertz While i understand empirical data, a lot of what happens here "feels" subjective, but accounting for confirmation bias, there's still quite a few test-users who do believe this to be of significant value. Also, researchers agree in published books we are currently reading regarding this layer of filtering.

@ledvinap the math there is fantastic. There's definitely a lot of opportunity for further optimizations :)

rs2k commented:
Hey guys, it's probably worth mentioning I used 32 KHz logging to refine this filter. I don't believe this is something BF is currently capable of, but I was able to log raw and output of one axis at a time at full 32 KHz operation in RF1. Here's an example csv file I that I put together and feed through C programs on the computer to check output with.

https://www.dropbox.com/s/83qebfl728ipgjj/32data.csv?dl=0

Also, I'm completely aware it's possible to further optimize this static filter, but I don't believe that's a very good use of time. I've used more than one dynamic version where the Kalman gain is constantly changing that show quite a bit more promise and will require this static filter to stay in the form it is to work correctly.

DanNixon commented:
@timothyswt Again, the same reasons I've already mentioned. Mainly non-determinism, you can't evaluate a filter if you can't supply it (and those existing for comparison) with the same sample data repeatedly. Also the fact it is much faster to run simulations than to implement filters, supply test firmware, wait for pilots to do test flights then interpret their logs.

I've not seen anything regarding @denovich 's method or results, only a statement. Statements alone carry no weight.

conkerkh commented:
I can log at 32k most likely, but I need to fix up my SDIO commit there are some issues on multiblockwrite. Once that's done I'll try to get you blackbox logged at 32K

rs2k commented:
@DanNixon @conkerkh I just supplied raw and filtered yaw gyro output in CSV form if you want to play with that. This is from a hovering quad running full 32/32/32 with some stick inputs on the yaw axis. I've found this is the best way to start tests on something new. I've collected many many logs like this for testing and it's amazing how much of a difference the filter makes in flight compared to biquad and PT1. While you can see a difference during hovers, you really start seeing the difference once gyro clipping starts happening in hard flying conditions. That simply something you can't test for using MATLAB.

timothyswt commented:
@DanNixon You're basically saying, "I don't want to spend the time to do it correctly, so I'm going to do it the way I want and call it 'science.'" The simplest method is not always the most correct or even yield the correct results.

Also, if statements carry no weight, why are you even attempting to complain? You're actually asserting "Only my statements carry weight."

DanNixon commented:
@rs2k I would typically simulate using captured raw data, not synthetic data alone.

@timothyswt 

> You're actually asserting "Only my statements carry weight."

No. I'm asserting that unfounded statements carry no weight.

timothyswt commented:
@DanNixon 

> No. I'm asserting that unfounded statements carry no weight.

Either you're trolling or just literally refusing to read the other messages and look at the data.
![image](https://user-images.githubusercontent.com/1736223/34783729-906b810a-f5e9-11e7-83c4-7c606c54ccfa.png)

wind0r commented:
@timothyswt Where is this data?

timothyswt commented:
@wind0r look at the comments made by @rs2k
He posts links to the test data used to test the filter.

rs2k commented:
@DanNixon simulations are good for starting, but they don't hold a candle to real life testing. There's a lot of subtlety in the captured data that feeds back onto itself in ways that can't be simulated with current technology. How does MATLAB handle gyro clipping? How does MATLAB simulate electrical noise generated by changes in the filters output that end up being fed back into the ESCs which in turn create different noise? How does MATLAB handle a fluttering prop that stalls during a hard 180? 

We're doing more than just signal processing here. We're not trying to look for a signal from space that's not affected by our process, we're trying to look for a digital signal that's a representation of an analog signal of a mechanical device that's affected by electrical noise, hardware vibrations, and even sound vibrations... all of which are affected by how you read and interpret the signal. MATLAB can't even come close to simulating that.

DanNixon commented:
@rs2k I agree, I never said simulation is a complete replacement for actual testing. However it is IMO the most reliable way to draw non-subjective comparisons between differing approaches.

I accept the point on how feedback can effect the sample data, however you also have this problem with test flights given the large permutation of hardware giving each potentially giving different responses. Unless you make a general assumption about the introduced noise.

As a side note, do you have any data that shows this noise being introduced? (asking out of personal interest, not to try and make a point)

ledvinap commented:
@rs2k : There is surprisingly little magic related to filtering. There is Nyquist theorem, there is causality requirement; pulse response for FIR filters and poles/zeros for IIR.
For example there is always direct relationship between filter frequency response and filter delay. It doesn't matter much how the filter is implemented, be it mechanical, analog or digital. For given amount of attenuation, you get about identical delay.
Small exception here are notch filters, gaining some improvement by not filtering high frequencies, and FIR filters with total attenuation on some (discrete) frequencies.

Applying Kalman to gyro data is possible, but inefficient - filter will quickly converge to simple first order IIR (by definition of what Kalman filter is). There may be some space for gain scheduling, but IMO gyro filter is wrong place to apply it.

Using tools/algorithms with little understanding how they work does work in some cases, but in my experience it quickly ends in huge mess, mixing layer over layer of heuristics, producing code that is extremely hard to understand and very fragile.

ledvinap commented:
BTW: here is Octave code to test filter equivalence:
``` octave
len = 1000;

lastX = 0;
q = 1600 * 0.000001;
r = 1600 * 0.001;

vx = zeros([len,1]);
vk = zeros([len,1]);
input = rand([len,1]);

x = 0;
p = 0;

for i = 1:length(input)
  x += (x - lastX);

  % update last state
  lastX = x;

  % prediction update
  p = p + q;

  % measurement update
  k = p / (p + r);
  x += k * (input(i) - x);
  p = (1 - k) * p;

  vx(i)=x;
  vk(i)=k;
endfor

% 1st order IIR
k = 2 * 2 * sqrt(q) / (sqrt(q + 4 * r) + sqrt(q));  % RC filter k
b = [k/2, k/2, 0];
a = [1, -(1-k), 0];

fx = filter(b,a,input);   % IIR filter input data
plot(fx-vx);

timothyswt commented:
Guys, honestly, if you don’t want the kalman filtering don’t take it. RF1 in open source will continue to advance beyond all of this anyways. If you do want it, merge it. Otherwise, it’s a waste of time arguing over a clear benefit of having this as another filtering option within BF. You could even consider it an experimental option for now and then, when you’ve decided you’ve “proven” @rs2k “wrong” (which is dubious) remove it.

mikeller commented:
@timothyswt: I don't think you understand what this is about (by the way, the 'kalman' filter pull request has been merged already). We are simply trying to understand what this filter actually does, and how it potentially can be made more efficient / integrated more seamlessly. Doing this also helps to assess it based on its merits, and not based on (potentially hyped up) empirical data.

ledvinap commented:
@rs2k : The data (https://www.dropbox.com/s/83qebfl728ipgjj/32data.csv?dl=0) seems to be samples with Ts=3ms. Is that correct or is there some scaling involved?

ctzsnooze commented:
I recall the work that went into evaluating the PT1, Biquad and FIR filters that have been the basis for filtering in betalfight these last few years.  We were able to predict very readily which worked better when we had graphs showing frequency and phase characteristics of the various types of filters under different circumstances. 

When proposing new filters, we need at least to know the frequency response and phase characteristics of the new / proposed filter, compared to what we already have.  

One way this can be done is by running real recorded data through a simulation of each filter and comparing pre- and post- filter spectrums.  e.g. unfiltered spectrum vs PT1 vs biquad vs Kalman vs Biquad+FIR2.

The other way, and the way we did it before, was for someone to graph, using Opera or something, frequency and phase response of each filter.  

It's great to know that pilots like a new filter, but to ensure we have the best solution we need to also see what the filter is actually doing to the data.  Then we can figure out *why* the code being flown feels better.  It could be that (say) the Kalman is only doing the same as a PT1, but just that the numbers being input are resulting is cutoff at 150Hz.  That would feel better than at 100Hz.  Or it could be that Kalman results in less delay, or steeper cut, or whatever.  All of which helps us understand how best to implement it and to evaluate alternative methods of achieving the same.

So yeah I am 100% in agreement that we must evaluate any proposed filter with proper data showing what it does.  

IMHO we should not implement any code until we know, precisely and technically, exactly what the code is doing.  We knew this for PT1 and Biquad and FIR, mostly thanks to Ledvinap.

If we don't know what some code actually does, how can we explain to users what it does - let alone how can we best explain how and when to use it ???

ctzsnooze commented:
This is not about proving anyone right or wrong.  Its about scientifically and objectively evaluating a suggestion that may be a big step forward to determine how it actually does its magic.

DieHertz commented:
Love how impartially and professionally you have summed it up, @ctzsnooze :+1:
